pkg/kubelet/dockertools/manager.go

```diff
@@ -1769,6 +1756,73 @@ func (dm *DockerManager) computePodContainerChanges(pod *api.Pod, podStatus *kub
 	}, nil
 }
 
+// tearDownPodNetwork calls the network plugin to tear down the pod (if
+// applicable) and delete the infra container.
+func (dm *DockerManager) tearDownPodNetwork(pod *api.Pod, runningPod *kubecontainer.Pod, networkSpec *api.Container, container *kubecontainer.Container) error {
+	ins, err := dm.client.InspectContainer(container.ID.ID)
+	if err != nil {
+		return fmt.Errorf("Error inspecting container %v: %v", container.ID.ID, err)
+	}
+	if getDockerNetworkMode(ins) != namespaceModeHost {
+		// pod could be nil, so we use the runningPod for name and namespace.
+		if err := dm.networkPlugin.TearDownPod(runningPod.Namespace, runningPod.Name, kubecontainer.DockerID(container.ID.ID)); err != nil {
+			return fmt.Errorf("Failed to teardown network for pod %q using network plugins %q: %v", pod.UID, dm.networkPlugin.Name(), err)
+		}
+	}
+	if err := dm.KillContainerInPod(container.ID, networkSpec, pod, "Need to kill pod."); err != nil {
+		return err
+	}
+	return nil
+}
+
+// setUpPodNetwork creates an infra container and calls the network plugin to
+// join the network if necessary. It returns the container ID of the infra
+// container and the pod IP.
+func (dm *DockerManager) setUpPodNetwork(pod *api.Pod) (kubecontainer.DockerID, string, error) {
+	glog.V(4).Infof("Creating pod infra container for %q", format.Pod(pod))
+	containerID, err, _ := dm.createPodInfraContainer(pod)
+	if err != nil {
+		glog.Errorf("Failed to create pod infra container: %v; Skipping pod %q", err, format.Pod(pod))
+		return containerID, "", err
+	}
+
+	if usesHostNetwork(pod) {
+		return containerID, "", nil
+	}
+
+	// Call the networking plugin
+	err = dm.networkPlugin.SetUpPod(pod.Namespace, pod.Name, containerID)
+	if err != nil {
+		// TODO: (random-liu) There shouldn't be "Skipping pod" in sync result message
+		message := fmt.Sprintf("Failed to setup network for pod %q using network plugins %q: %v; Skipping pod", format.Pod(pod), dm.networkPlugin.Name(), err)
+		glog.Error(message)
+
+		// Delete the infra container we just created.
+		if delErr := dm.KillContainerInPod(kubecontainer.ContainerID{
+			ID:   string(containerID),
+			Type: "docker"}, nil, pod, message); delErr != nil {
+			glog.Warningf("Clear infra container failed for pod %q: %v", format.Pod(pod), delErr)
+			err = fmt.Errorf("%v; failed to delete the infra container %v", err, delErr)
+		}
+		return containerID, "", err
```

Comment:
Shouldn't containerID get cleared here, since the container was torn down and no longer exists?